designs/ui-module/README.md

```diff
@@ -0,0 +1,261 @@
+Layout Tabs & Lynks editor:  
+![layout-tab-link](./layout-tab-link.png)   
+*UI_SUBGROUP_I instance settings can be edited with the [edit] button on this screen.  
+To edit the layout of UI module(Template), open the UI_SUBGROUP_T template edit screen from the instance screen and press the [Edit Subgroup Layout] button to open the layout screen.
+
+### UI module function operation procedure
+
+The operation procedure when the user arranges the UI module on the dashboard is shown below.  
+
+1. Create SUBFLOW template  
+The user creates SUBFLOW template and places widget.
+
+2. Edit UI_SUBGROUP_T template  
+In step 1, Node-RED dashboard automatically creates UI_SUBGROUP_T template.  
+The user can edit the layout of UI module by opening the edit screen of UI_SUBGROUP_T template node from the node "info" tab.  
+
+3. Create SUBFLOW instance  
+Place SUBFLOW instance on tab or on SUBFLOW template.
+
+4. Edit UI_SUBGROUP_I instance  
+In step 3, Node-RED dashboard automatically generates UI_SUBGROUP_I instance.  
+The user opens the edit screen of UI_SUBGROUP_I instance node from the node "info" tab, etc., and specifies UI_GROUP of arbitrary Node-RED dashboard.  
+However, if you place it on SUBFLOW template, you do not need to specify UI_GROUP.
+
+5. Layout of dashboard  
+The user edits the layout of dashboard that contains the UI module.  
+UI module size is fixed.  
+
+6. SUBFLOW property support(optional)  
+If UI_GROUP is defined as the environment variable for the property definition of SUBFLOW, UI_GROUP of Node-RED dashboard can be specified in step 3.  
+In this case, the procedure for setting UI_GROUP in step 4 can be omitted. (See below)
+
+### UI module sharing
+
+UI module can be exported as SUBFLOW, which can be imported.  
+By setting the node ID of SUBFLOW template to the 'z' key, the SUBFLOW template that contains the "UI module configuration node" will be the export target.
+
+- UI_SUBGROUP_T template node  
+Set SUBFLOW template node to the 'z' key.
+
+- UI_SUBGROUP_I instance node  
+When arranging to SUBFLOW template: Set the ID of SUBFLOW template node to the 'z' key.  
+When arranging in TAB: 'z' key is not specified (set as Global Configlation Node).
+
+### Specification of UI_GROUP by SUBFLOW environment variable (optional function)
+
+By specifying UI_GROUP config node using the property editing of SUBFLOW, 
+you can specify the group that displays the UI component when placing SUBFLOW instance like normal widget.  
+
+Property name : `_UI_GROUP`
+
+UI_SUBGROUP_I instance node sets the value specified in the corresponding SUBFLOW instance _UI_GROUP property to the value specified in group.
+
+### Add features for UI modules to Node-RED
+
+To support UI module function, Node-RED must have the following additions.
+
+1. Acquisition of subflow information
+
+  - Additional functions  
+   Server-side RED.nodes.getNode () supports getting subflow instance information.
```

Comment:
Exactly what information is needed from the subflow instance?

 Is it the subflow env properties? If so, they will already be visible to the node as environment variables - `this._flow.getSetting("...")` can be used to access its value in the subflow.

Reply:
I want to get the ID information of the subflow instance to which the widget being generated belongs.
For widgets nested within a subflow, IDs are required up to the parent subflow.
Specifically, this is the value set for the subflow path.

Example:
 "fda95d7e.fb54/3b5a5523.01d41a/d771a7ef.6586e8/cc844746.a516f8"

In node-red-dashbord ui.js#function add(opt), I want to use the above value when arranging the widget, so I considered the idea of getting the subflow instance from the value of opt.node.z.
I checked the argument opt, but it seems that the '._flow' you pointed out does not exist.

Reply:
I would appreciate it if you could consider this matter.
>Reference implementation: [get subflow info](https://github.com/node-red-hitachi/node-red/commit/e29eca6c000844c6a044f0c2a0f89a2dbf18feed)

I recognize that it is necessary to make corrections based on this measure in order to acquire subflow information in the process of UI node generation.

The following usage is assumed.

node-red-dashboard#ui.js:
>function add(opt) {  
>   var subflow = nodes.getNode(opt.node.z);